### DISCUSSION TITLE
Improved sound support

### DISCUSSION THREAD

**Discussion by rsta2** [rsta2/circle#291] [Development]
My suggested modifications to the sound API are on the branch *sound-improved*. I moved all sound related classes to *include/circle/sound/* and *lib/sound/*. Please note that I used the term "sound" instead of "audio", when I started with adding sound drivers to Circle. This cannot be changed easily and should remain consistent.

There is a new class `CSoundController`, which represents an entity, which provides functions to control an audio stream, which is maintained by another class (e.g. `CI2SSoundBaseDevice`). Hardware specific implementations of `CSoundController` (e.g. `CPCM512xSoundController`) act on the I2C interface of the device and also provide a method for probing the existence of this device. The current declaration of `CSoundController` is only an example, and should be extended in several ways.

The class `CSoundBaseDevice` got an additional virtual method `GetController()`, which returns a pointer to the sound controller for the respective class. Please note that `GetController()` can return `nullptr`, when there is no controller available (e.g. for `CPWMSoundBaseDevice` or for PCM5102A I2S devices).

As an example, I added a volume control via the MIDI CC event 7 to *sample/29-miniorgan/*. This is currently implemented and tested for PCM512x devices only. There is also a sound controller implemented for the class `CVCHIQSoundBaseDevice` in *addon/vc4/sound/*.

What do you think about this?

**Comment by rsta2** [comment]
If somebody wants to contribute support for additional I2S-based sound devices, you are invited to send a pull request to be merged into the branch *sound-improved*. If you want to support more features than the currently provided in the class `CSoundController`, you can extend this class declaration by more virtual methods. Let me know, when you have questions!

**Comment by sebastienNEC** [comment]
Hello René, just a quick message to say thank you for the proposal and sorry for the late answer (still swamped lately). I should be able at some stage to contribute controllers for the FePi, The PiSound, and the MAX98090 codecs/boards. I'll probably need to modify the input/output selector API in the case where the user would like to use **both** line output and headphone output for example on the MAX98090. And probably to add things like jack detection. All in due time :)

**Reply by rsta2** [reply]
Hi Sebastien, it would be great to have support for the codecs/boards, you have noted. There is no hurry. I expect Circle 44.5 to come at the end of this month, but we can easily plan to add this to the next following release. Feel free to make the modifications to the `CSoundController` interface, that you need.

**Reply by rsta2** [reply]
@sebastienNEC Have you made some progress in this field? I'm thinking about merging the *sound-improved* branch into *develop* for the next release already, but this makes only sense, if the class `CSoundController` will be implemented for some sound devices now or later, because users have to modify their applications to include the sound classes from a new directory. There should be some gain for these costs. I also thought about, if some class `CSoundJack` is necessary to represent different outputs and inputs of a sound card. But honestly I'm not familiar enough with the different, more complex sound cards available, to decide this on my own.

**Reply by sebastienNEC** [reply]
Hello Rene, I am very sorry that I didn't manage to work on this at all. Intuitively, I would say that merging this is a bit premature because of the above-mentioned limitations (only one output/input selector, no separate left/right volume settings, etc).
Unless of course somebody else needs this support, but if I am the only "client", I would elect to polish this a bit more until merging.
What do you think ?

**Reply by rsta2** [reply]
Hi Sebastien, are you ready to deal with this in the next weeks? Then I would give you the time to implement your modifications. Otherwise I would continue to work on the USB audio class support, which requires support from the (improved) sound system too. I guess it's not a good idea, if two developers are working on the same code at the same time. I think the next Circle release should come with an enhancement for the sound system in December. It would be great, if you could contribute your code to this.

**Comment by sebastienNEC** [comment]
Hello Rene, I am afraid I will not be able to work on this in the next few weeks :-/

**Reply by rsta2** [reply]
Hi Sebastien, that's a pity but perhaps there is another way, because there should be a new Circle release in foreseeable time. I can merge the *sound-improved* branch to *develop* now and continue to work on it to improve the USB audio support, I am working on. I have an USB sound device here, which has multiple jacks and also needs an API, which supports this. I'll try to implement some preliminary interface, which can be modified later. I will point out in the release notes, that each service, which is accessible via the class `CSoundController` may be a subject of future modifications. So when you have the time to implement the specific drivers for the sound devices, you own, you will be free to do the necessary updates.

**Reply by sebastienNEC** [reply]
Yes indeed, this feels like the most efficient way forward. Thank you much and sorry again for the inactivity. I'll get back to this project in due time :)

**Reply by rsta2** [reply]
Thanks for agreeing. I have merged the *sound-improved* branch to *develop* (see below). I'm looking forward to your valuable contributions, when you have some time!

**Comment by rsta2** [comment]
The branch *sound-improved* has been merged to *develop*. *sound-improved* will be removed. For users, who work with the *develop* branch, this has one important effect: All classes `C*Sound*Device` are implemented in *lib/sound/* now. The library *lib/sound/libsound.a* must be added to the list of libraries, if one of the sound classes is used in your application. The respective header files are in *include/circle/sound/* now. Some `#include` directives may have to be updated in your application.

**Comment by rsta2** [comment]
Circle 45 is out with some improvements for the sound support. This is probably not the end of the development in this direction. Users are invited to contribute code for new codecs and features.

**Comment by sebastienNEC** [comment]
Great news thank you ! The first codec I'll implement is the venerable wm8731 - I need it relatively quickly for a project of mine :)

**Reply by rsta2** [reply]
Great! Let me know, when you need help.
